### PR TITLE
Using order status constants in main files

### PR DESCRIPTION
**Pull request by wjrosa**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Base PR https://github.com/woocommerce/woocommerce/pull/51919.

This PR uses the constants introduced in https://github.com/woocommerce/woocommerce/pull/51919 on the main files of all woocommerce plugin codebase. Previous PRs replaced constant usages only in unit tests. This is the first actually using it on live code.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

- Checkout and build this branch on your test environment (`dev/using-order-status-constants-in-the-main-files`)
- Perform some smoke tests such as:
    - Store setup
    - Payment gateway installation (Stripe, for example)
    - Product creation
    - Common checkout
    - Shortcode checkout
    - My account page navigation
    - Order listing/details
- Confirm that no regression was introduced

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Uses the newly introduced order status constants in the main WooCommerce plugin codebase.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by kalessil — CHANGES_REQUESTED** [review]
Nice! 

@wjrosa: some of the CI checks are failing. Can you take a look, please?

**Comment by wjrosa** [comment]
> Nice!
> 
> @wjrosa: some of the CI checks are failing. Can you take a look, please?

Thanks, @kalessil ! All fixed ✅

**Comment by kalessil** [comment]
@barryhughes: this PR is touching areas covered by multiple teams; what would be our take here (split in smaller PRs or keep the current PR size)?

**Comment by barryhughes** [comment]
I'm not sure, PRs of this nature are always a little tricky, and the 'dividing lines' between teams can be pretty blurry. My hunch is to keep it as a single unit on this occasion.

Noting also there seem to be fresh CI fails since the last commit.

**Comment by wjrosa** [comment]
> Noting also there seem to be fresh CI fails since the last commit.

It looks like it was not related to this PR itself. After updating the PR with the base branch and re-running the checks, they passed 👍

File: plugins/woocommerce/includes/abstracts/abstract-wc-product.php
Comment (barryhughes) [inline plugins/woocommerce/includes/abstracts/abstract-wc-product.php]: This shouldn't cause problems, because `OrderStatus::TRASH === 'trash'`, but it still doesn't feel right to use the order status here within the definition of the product type (ie, product statuses are not order statuses, even if there is overlap in terms of the actual status names we use).

File: plugins/woocommerce/includes/admin/wc-admin-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/admin/wc-admin-functions.php]: Similar to an earlier note, this works, but I'm not sure if it's appropriate as we're examining post statuses (for pages), rather than order statuses.

File: plugins/woocommerce/includes/class-wc-cart.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-cart.php]: Same question re conflating statuses (product vs order).

File: plugins/woocommerce/includes/class-wc-cart.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-cart.php]: Here also ... same question re conflating statuses (product vs order).

File: plugins/woocommerce/includes/class-wc-post-data.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-post-data.php]: One more to check on (order vs product).

File: plugins/woocommerce/includes/class-wc-shortcodes.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-shortcodes.php]: Here also, I think we're interested in the product (post) status, not the order status.

File: plugins/woocommerce/includes/class-wc-webhook.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-webhook.php]: Here also, I think perhaps we are interested in products and coupons, rather than orders.

File: plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php
Comment (barryhughes) [inline plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:108]: If `checkout-draft` is a possibility, should it be added to the enumeration class? Not immediately sure about this, but wanted to pause and ask.

File: plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php
Comment (barryhughes) [inline plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:220]: Depending on our decision to the earlier question re `checkout-draft`, we *might* want to update this line.

File: plugins/woocommerce/includes/wc-product-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-product-functions.php]: Product/post status rather than an order status?

**Review by barryhughes — COMMENTED** [review]
Left a few questions, mostly checking to see if the use of `OrderStatus` is appropriate in some cases (where we are dealing with posts, pages, products or coupons, etc).

**Comment by wjrosa** [comment]
Thanks for the detailed look, @barryhughes ! I have reverted all the invalid replacements you flagged. I was afraid of doing those, but sometimes it is hard to be sure of the context. The additional eyes are very welcome.

As for the `checkout-draft,` @kalessil suggested leaving those for the future since this constant belongs to `DraftOrders::STATUS.`

**Comment by barryhughes** [comment]
> As for the checkout-draft, @kalessil suggested leaving those for the future since this constant belongs to DraftOrders::STATUS.

Great (and apologies for missing that) 👍🏼

**Review by barryhughes — APPROVED** [review]
Nothing else to add, thanks! Not sure if you need or want other folks from the list of reviewers to chime in or not but, if you don't, I guess we can go ahead and merge ✅

**Comment by wjrosa** [comment]
I think now I need another review from @kalessil since he requested changes earlier

File: plugins/woocommerce/includes/wc-product-functions.php
Comment (kalessil) [inline plugins/woocommerce/includes/wc-product-functions.php]: Unused import: we should drop it before merging.

File: plugins/woocommerce/includes/abstracts/abstract-wc-product.php
Comment (kalessil) [inline plugins/woocommerce/includes/abstracts/abstract-wc-product.php]: Unused import.

File: plugins/woocommerce/includes/class-wc-shortcodes.php
Comment (kalessil) [inline plugins/woocommerce/includes/class-wc-shortcodes.php]: Unused import.

File: plugins/woocommerce/includes/class-wc-cart.php
Comment (kalessil) [inline plugins/woocommerce/includes/class-wc-cart.php]: Unused import.

File: plugins/woocommerce/includes/gateways/cheque/class-wc-gateway-cheque.php
Comment (kalessil) [inline plugins/woocommerce/includes/gateways/cheque/class-wc-gateway-cheque.php]: This way, the filter invocation is a bit preliminary, so it might be more invocations than necessary.

I'm suggesting refining approach as follows to address that:
```
if ( $this->instructions && ! $sent_to_admin && 'cheque' === $order->get_payment_method() ) {
   // doc-block here
   $instructions_order_status = apply_filters( 'woocommerce_cheque_email_instructions_order_status', OrderStatus::ON_HOLD, $order );
    if ( $order->has_status( $instructions_order_status )  ) {
        echo wp_kses_post( wpautop( wptexturize( $this->instructions ) ) . PHP_EOL );
    }
}
```

Comment (kalessil) [reply plugins/woocommerce/includes/gateways/cheque/class-wc-gateway-cheque.php]: It might be more similar patterns in the PR, but let's align on this one first.

**Review by kalessil — CHANGES_REQUESTED** [review]
We almost there, @wjrosa!

Leaving some comments to look into:
- there are some unused imports we can drop
- I spot a case where we change filters invocation frequency, it can be more of them

File: plugins/woocommerce/includes/wc-product-functions.php
Comment (wjrosa) [reply plugins/woocommerce/includes/wc-product-functions.php]: Good catch, Vladimir! I have addressed all the unused import cases.

File: plugins/woocommerce/includes/gateways/cheque/class-wc-gateway-cheque.php
Comment (wjrosa) [reply plugins/woocommerce/includes/gateways/cheque/class-wc-gateway-cheque.php]: Makes sense @kalessil . I have addressed it in https://github.com/woocommerce/woocommerce/pull/51922/commits/af592ec50cc9e35e0a3dc9d6e884b7cad922e40f

File: plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php
Comment (kalessil) [inline plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php]: Filter invocation frequency change.

File: plugins/woocommerce/includes/gateways/bacs/class-wc-gateway-bacs.php
Comment (kalessil) [inline plugins/woocommerce/includes/gateways/bacs/class-wc-gateway-bacs.php]: Filter invocation frequency change.

**Review by kalessil — APPROVED** [review]
I'm leaving a couple of comments to address and pre-approving; great work here @wjrosa  !

File: plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php
Comment (wjrosa) [reply plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php]: Fixed in https://github.com/woocommerce/woocommerce/pull/51922/commits/9071d081dfe753615780eec18284c19023d255c9

File: plugins/woocommerce/includes/gateways/bacs/class-wc-gateway-bacs.php
Comment (wjrosa) [reply plugins/woocommerce/includes/gateways/bacs/class-wc-gateway-bacs.php]: Fixed in https://github.com/woocommerce/woocommerce/pull/51922/commits/9071d081dfe753615780eec18284c19023d255c9

**Pull request merged**
